crates/goose/src/agents/agent.rs

```
@@ -173,6 +176,89 @@ impl Agent {
+        *retry_attempts += 1;
+        *retry_attempts
+    }
+
+    /// Get the current retry attempts count
+    pub async fn get_retry_attempts(&self) -> u32 {
+        *self.retry_attempts.lock().await
+    }
+
+    /// Handle retry logic for the agent reply loop
+    async fn handle_retry_logic(
+        &self,
+        messages: &mut Vec<Message>,
+        session: &Option<SessionConfig>,
+        initial_messages: &[Message],
+    ) -> Result<bool> {
+        let Some(session_config) = session else {
+            return Ok(false); // No retry if no session config
+        };
+
+        let Some(retry_config) = &session_config.retry_config else {
+            return Ok(false); // No retry if no retry config
+        };
+
+        // Execute success checks
+        let success = execute_success_checks(&retry_config.checks, retry_config).await?;
+
+        if success {
+            info!("All success checks passed, no retry needed");
+            return Ok(false); // Success, no retry needed
+        }
+
+        // Check retry attempts
+        let current_attempts = self.get_retry_attempts().await;
+        if current_attempts >= retry_config.max_retries {
+            // Add max retries exceeded message
+            let error_msg = Message::assistant().with_text(format!(
+                "Maximum retry attempts ({}) exceeded. Unable to complete the task successfully.",
+                retry_config.max_retries
+            ));
+            messages.push(error_msg);
+            warn!(
+                "Maximum retry attempts ({}) exceeded",
+                retry_config.max_retries
+            );
+            return Ok(false); // No more retries
+        }
+
+        // Execute cleanup command if provided
+        if let Some(cleanup_cmd) = &retry_config.on_failure {
+            info!("Executing cleanup command: {}", cleanup_cmd);
+            if let Err(e) = execute_cleanup_command(cleanup_cmd, retry_config).await {
```

I'm wondering, if someone has gone to the trouble of defining an on_failure fn (probably better to name the fn the same as the config parameter) it means they are expecting a particular state to be in play. So if this fails I feel like we should let the Err bubble up so the agent loop will end rather than still retrying in an unexpected state.

---

Yep, makes sense. I am going to let it bubble up.